pkg/scheduler/plugins/prioritize/prioritize.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prioritize
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/priorities"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	"k8s.io/kubernetes/pkg/scheduler/cache"
+
+	"github.com/kubernetes-sigs/kube-batch/pkg/scheduler/api"
+	"github.com/kubernetes-sigs/kube-batch/pkg/scheduler/framework"
+)
+
+type prioritizePlugin struct {
+}
+
+func getInterPodAffinityScore(name string, interPodAffinityScore schedulerapi.HostPriorityList) int {
+	for _, hostPriority := range interPodAffinityScore {
+		if hostPriority.Host == name {
+			return hostPriority.Score
+		}
+	}
+	return 0
+}
+
+func generateNodeMapAndSlice(nodes []*api.NodeInfo) (map[string]*cache.NodeInfo, []*v1.Node) {
+	var nodeMap map[string]*cache.NodeInfo
+	var nodeSlice []*v1.Node
+	nodeMap = make(map[string]*cache.NodeInfo)
+	for _, node := range nodes {
+		nodeInfo := cache.NewNodeInfo(node.Pods()...)
+		nodeInfo.SetNode(node.Node)
+		nodeMap[node.Name] = nodeInfo
+		nodeSlice = append(nodeSlice, node.Node)
+	}
+	return nodeMap, nodeSlice
+}
+
+type cachedNodeInfo struct {
+	session *framework.Session
+}
+
+func (c *cachedNodeInfo) GetNodeInfo(name string) (*v1.Node, error) {
+	node, found := c.session.NodeIndex[name]
+	if !found {
+		for _, cacheNode := range c.session.Nodes {
+			pods := cacheNode.Pods()
+			for _, pod := range pods {
+				if pod.Spec.NodeName == "" {
+					return cacheNode.Node, nil
+				}
+			}
+		}
+		return nil, fmt.Errorf("failed to find node <%s>", name)
+	}
+
+	return node.Node, nil
+}
+
+type podLister struct {
+	session *framework.Session
+}
+
+func (pl *podLister) List(selector labels.Selector) ([]*v1.Pod, error) {
+	var pods []*v1.Pod
+	for _, job := range pl.session.Jobs {
+		for status, tasks := range job.TaskStatusIndex {
+			if !api.AllocatedStatus(status) {
+				continue
+			}
+
+			for _, task := range tasks {
+				if selector.Matches(labels.Set(task.Pod.Labels)) {
+					pod := task.Pod.DeepCopy()
```

Comment:
we only need DeepCopy the pod that `task.NodeName != Pod.NodeName`.

Reply:
In interPodAffinity Priority code, we do not use PodLister's `List` and `FilteredList` function, list is done implicitly in the interPodAffinity code, where is ranges over every node and get pods from nodeinfo object, and calls processPod function, which uses only `GetNodeInfo` function alone.

```
processNode := func(i int) {
		nodeInfo := nodeNameToInfo[allNodeNames[i]]
                ......
		......		
               for _, existingPod := range nodeInfo.Pods() {
			if err := processPod(existingPod); err != nil {
```
  
In interPodAffinity Priority code of K8s, uses 
`existingPodNode, err := ipa.info.GetNodeInfo(existingPod.Spec.NodeName)`
function, since `existingPod.Spec.NodeName` value was nil, we were getting error, so have made changes only in `func (c *cachedNodeInfo) GetNodeInfo(name string) (*v1.Node, error)` function alone, other listers are same as there in Predicate plugin.  Changes has been done only to `GetNodeInfo` function for fixing that case.  Not in `FilteredList` and `List`, those are same as that of present in predicate plugin.  `FilteredList` and `List` function is not used in interPodAffinity priority code, it is defined only because that struct should implement a interface.  Changes in those function will not affect the behaviour of priority function.

Reply:
if `task.NodeName == pod.NodeName`, append pod as before;  if `task.NodeName != pod.NodeName`, do DeepCopy and then append copied pod. Do you mean that does not work?

Reply:
That `List` function where you are asking me to change, is not used by interPodAffinity priority code.

Reply:
what's different? If NodeName not change, why do we do DeepCopy? priority and predicate only read pod info.

Reply:
Have changed the implementation, according to your comments